### PR TITLE
Initial version for H265 decoder.

### PR DESCRIPTION
**Pull request by YuJiankang**

### CODE REVIEW

File: decoder/vaapidecoder_h265.cpp [hunk 1]
```diff
@@ -0,0 +1,1784 @@
...
+    VAAPI_DECODER_UNIT_FLAG_FRAME_END   = (1 << 1),
+    VAAPI_DECODER_UNIT_FLAG_STREAM_END  = (1 << 2),
+    VAAPI_DECODER_UNIT_FLAG_SLICE       = (1 << 3),
+    VAAPI_DECODER_UNIT_FLAG_SKIP        = (1 << 4),
+    VAAPI_DECODER_UNIT_FLAG_LAST        = (1 << 5)
+} VaapiDecoderUnitFlags;
+
+enum
+{
+  VAAPI_DECODER_UNIT_FLAG_AU_START =
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h265.cpp]: 4 spaces indent is better

File: decoder/vaapidecoder_h265.cpp [hunk 2]
```diff
@@ -0,0 +1,1784 @@
...
+}
+
+VaapiH265FrameStore::~VaapiH265FrameStore()
+{
+}
+
+VaapiDecoderH265::VaapiDecoderH265()
+{
+    m_flags = 0;
+    m_dpb_count = 0;
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h265.cpp]: better put to constructor init list

File: decoder/vaapidecoder_h265.cpp [hunk 3]
```diff
@@ -0,0 +1,1784 @@
...
+
+VaapiDecoderH265::~VaapiDecoderH265()
+{
+    stop();
+}
+Decode_Status VaapiDecoderH265::start(VideoConfigBuffer *buffer)
+{
+    Decode_Status status;
+    buffer->profile = VAProfileHEVCMain;
+    buffer->surfaceNumber = 30;
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h265.cpp]: better choose surfaceNumber base on reference list

File: decoder/vaapidecoder_h265.cpp [hunk 4]
```diff
@@ -0,0 +1,1784 @@
...
+    return nal_header_bytes + (slice_hdr->header_size + 7) / 8 - epb_count;
+}
+
+bool VaapiDecoderH265::fillPredWeightTable(VASliceParameterBufferHEVC*sliceParam)
+{
+    uint32_t i, j;
+    sliceParam->luma_log2_weight_denom = 0;
+    sliceParam->delta_chroma_log2_weight_denom = 0;
+
+    if (0) {
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h265.cpp]: so, we do not support weight table yet?

File: decoder/vaapidecoder_h265.h [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    virtual Decode_Status decode(VideoDecodeBuffer *);
+
+  private:
+    Decode_Status ensureContext();
+    Decode_Status decodeNalu(H265NalUnit * nalu);
+    uint32_t get_slice_data_byte_offset (H265SliceHdr * slice_hdr, uint32_t nal_header_bytes);
+    bool fillPredWeightTable(VASliceParameterBufferHEVC*sliceParam);
+    bool fillSlice(VaapiDecPictureH265 * picture, VASliceParameterBufferHEVC* sliceParam, H265SliceHdr * sliceHdr, H265NalUnit * nalu);
+    bool fillPicture(PicturePtr& picture, H265NalUnit * nalu);
+    void fill_iq_matrix_4x4 (VAIQMatrixBufferHEVC * iq_matrix, H265ScalingList * scaling_list);
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h265.h]: better use Camel style name conversion

File: decoder/vaapidecoder_h265.cpp [hunk 6]
```diff
@@ -0,0 +1,1784 @@
...
+    COPY_FIELD (pps, init_qp_minus26);
+    COPY_FIELD (pps, diff_cu_qp_delta_depth);
+    pic_param->pps_cb_qp_offset = pps->cb_qp_offset;
+    pic_param->pps_cr_qp_offset = pps->cr_qp_offset;
+    COPY_FIELD (pps, log2_parallel_merge_level_minus2);
+    COPY_FIELD (pps, num_tile_columns_minus1);
+    COPY_FIELD (pps, num_tile_rows_minus1);
+    for (i = 0; i <= pps->num_tile_columns_minus1; i++)
+        pic_param->column_width_minus1[i] = pps->column_width_minus1[i];
+    for (; i < 19; i++)
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h265.cpp]: better use N_ELEMENT instead of magic numbber 19

Comment (xuguangxin) [inline decoder/vaapidecoder_h265.cpp]: and even no need set to zero since we memset all member in ->editXXX

File: decoder/vaapidecoder_h265.cpp [hunk 7]
```diff
@@ -0,0 +1,1784 @@
...
+
+    /*Fixme: Set correct value as mentioned in va_dec_hevc.h */
+    pic_param->st_rps_bits = 0;
+
+    return true;
+}
+
+void VaapiDecoderH265::fill_iq_matrix_4x4 (VAIQMatrixBufferHEVC * iq_matrix,
+    H265ScalingList * scaling_list)
+{
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h265.cpp]: 4x4, 8x8  16x16 32x32 all seems same function, can we use a common function, pass array and zigzag function pointer to it?

File: decoder/vaapidecoder_h265.cpp [hunk 8]
```diff
@@ -0,0 +1,1784 @@
...
+        (iq_matrix->ScalingList32x32[i], scaling_list->scaling_lists_32x32[i]);
+    }
+}
+
+void VaapiDecoderH265::fill_iq_matrix_dc_16x16 (VAIQMatrixBufferHEVC * iq_matrix,
+    H265ScalingList * scaling_list)
+{
+    uint32_t i;
+
+    for (i = 0; i < 6; i++)
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h265.cpp]: magic number and similar function as 32x32

File: decoder/vaapidecoder_h265.cpp [hunk 9]
```diff
@@ -0,0 +1,1784 @@
...
+  /* Fixme: Add limit check based on Annex A */
+  return MAX (1, (sps->max_dec_pic_buffering_minus1[0] + 1));
+}
+
+Decode_Status VaapiDecoderH265::ensure_dpb (H265SPS * sps)
+{
+    uint32_t dpb_size;
+    dpb_size = get_max_dec_frame_buffering (sps);
+    if (m_dpb_size < dpb_size) {
+    }
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h265.cpp]: why it's empty?

File: decoder/vaapidecoder_h265.cpp [hunk 10]
```diff
@@ -0,0 +1,1784 @@
...
+}
+
+
+/* Get the dpb picture having the specifed poc or poc_lsb */
+VaapiDecPictureH265 *VaapiDecoderH265::dpb_get_picture (int32_t poc, bool match_lsb)
+{
+    uint32_t i;
+
+    for (i = 0; i < m_dpb_count; i++) {
+    VaapiDecPictureH265 *picture = m_dpb[i]->buffer.get();
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h265.cpp]: indent false

File: decoder/vaapidecoder_h265.cpp [hunk 11]
```diff
@@ -0,0 +1,1784 @@
...
+void VaapiDecoderH265::vaapi_picture_h265_set_reference(VaapiDecPictureH265* picture, uint32_t referenceFlags)
+{
+    VAAPI_PICTURE_FLAG_UNSET(picture, VAAPI_PICTURE_FLAGS_RPS_ST | VAAPI_PICTURE_FLAGS_RPS_LT);
+    VAAPI_PICTURE_FLAG_UNSET(picture, VAAPI_PICTURE_FLAGS_REFERENCE);
+    VAAPI_PICTURE_FLAG_SET(picture, referenceFlags);
+}
+
+
+/* Get the dpb picture having the specifed poc or poc_lsb */
+VaapiDecPictureH265 *VaapiDecoderH265::dpb_get_picture (int32_t poc, bool match_lsb)
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h265.cpp]: Try use a nest class for all dpb.
You can refer to https://github.com/xuguangxin/libyami/blob/grid/examples/grid.cpp. Flipper is a nest class of DrmRenderer

File: decoder/vaapidecoder_h265.cpp [hunk 12]
```diff
@@ -0,0 +1,1784 @@
...
+                i++;
+        }
+    }
+}
+
+void VaapiDecoderH265::dpb_flush ()
+{
+    /* Output any frame remaining in DPB */
+    while (dpb_bump (NULL));
+    dpb_clear (true);
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h265.cpp]: false indent

File: decoder/vaapidecoder_h265.cpp [hunk 13]
```diff
@@ -0,0 +1,1784 @@
...
+    memset (m_RefPicSetLtFoll, 0, sizeof (VaapiDecPictureH265 *) * 16);
+    memset (m_RefPicSetStCurrBefore, 0, sizeof (VaapiDecPictureH265 *) * 16);
+    memset (m_RefPicSetStCurrAfter, 0, sizeof (VaapiDecPictureH265 *) * 16);
+    memset (m_RefPicSetStFoll, 0, sizeof (VaapiDecPictureH265 *) * 16);
+
+    /* (8-6) */
+    for (i = 0; i < m_NumPocLtCurr; i++) {
+        if (!CurrDeltaPocMsbPresentFlag[i]) {
+            dpb_pic = dpb_get_picture (m_PocLtCurr[i], true);
+            if (dpb_pic)
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h265.cpp]: what is the difference between this if and a simple statement "m_RefPicSetLtCurr[i]  = dpb_pic;"

File: decoder/vaapidecoder_h265.cpp [hunk 14]
```diff
@@ -0,0 +1,1784 @@
...
+    memset (m_RefPicSetLtCurr, 0, sizeof (VaapiDecPictureH265 *) * 16);
+    memset (m_RefPicSetLtFoll, 0, sizeof (VaapiDecPictureH265 *) * 16);
+    memset (m_RefPicSetStCurrBefore, 0, sizeof (VaapiDecPictureH265 *) * 16);
+    memset (m_RefPicSetStCurrAfter, 0, sizeof (VaapiDecPictureH265 *) * 16);
+    memset (m_RefPicSetStFoll, 0, sizeof (VaapiDecPictureH265 *) * 16);
+
+    /* (8-6) */
+    for (i = 0; i < m_NumPocLtCurr; i++) {
+        if (!CurrDeltaPocMsbPresentFlag[i]) {
+            dpb_pic = dpb_get_picture (m_PocLtCurr[i], true);
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h265.cpp]: why not use dpb_get_picture (m_PocLtCurr[i], !CurrDeltaPocMsbPresentFlag[i])?

File: decoder/vaapidecoder_h265.cpp [hunk 15]
```diff
@@ -0,0 +1,1784 @@
...
+            if (dpb_pic)
+                m_RefPicSetLtCurr[i] = dpb_pic;
+            else
+                m_RefPicSetLtCurr[i] = NULL;
+        }
+    }
+    for (; i < 16; i++)
+        m_RefPicSetLtCurr[i] = NULL;
+
+    for (i = 0; i < m_NumPocLtFoll; i++) {
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h265.cpp]: similar logical, can we use same sub function as previous?

File: decoder/vaapidecoder_h265.cpp [hunk 16]
```diff
@@ -0,0 +1,1784 @@
...
+
+    memset (m_RefPicList0, 0, sizeof (VaapiDecPictureH265 *) * 16);
+    memset (m_RefPicList1, 0, sizeof (VaapiDecPictureH265 *) * 16);
+    m_RefPicList0_count = m_RefPicList1_count = 0;
+
+    if ((slice_hdr->type != H265_B_SLICE) &&
+        (slice_hdr->type != H265_P_SLICE))
+        return;
+
+    if (slice_hdr->dependent_slice_segment_flag) {
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h265.cpp]: better if we can use this

H265SliceHdr* hdr;
if (slice_hdr->dependent_slice_segment_flag)
    hdr = m_prev_independent_slice;
else
   hdr = slice_hdr;
num_ref_idx_l0_active_minus1 = hdr->num_ref_idx_l0_active_minus1;
.....

File: decoder/vaapidecoder_h265.cpp [hunk 17]
```diff
@@ -0,0 +1,1784 @@
...
+
+    /* construct RefPicList0 (8-9) */
+    for (rIdx = 0; rIdx <= num_ref_idx_l0_active_minus1; rIdx++)
+        m_RefPicList0[rIdx] =
+        ref_pic_list_modification->ref_pic_list_modification_flag_l0 ?
+        RefPicListTemp0[ref_pic_list_modification->list_entry_l0[rIdx]] :
+        RefPicListTemp0[rIdx];
+    m_RefPicList0_count = rIdx;
+
+    if (type == H265_B_SLICE) {
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h265.cpp]: is any possible to merge this with P frame code?

File: tests/decodeinputavformat.cpp [hunk 18]
```diff
@@ -53,6 +53,8 @@ bool DecodeInputAvFormat::initInput(const char* fileName)
             if (ctx->extradata && ctx->extradata_size)
                 m_codecData.append((char*)ctx->extradata, ctx->extradata_size);
             m_videoId = i;
+            m_videoStreamInfo.width  =  ctx->width;
```
Comment (xuguangxin) [inline tests/decodeinputavformat.cpp]: why we need this, can we get it from sps or rps

**Comment by YuJiankang** [comment]
This submission only used for the first round of code review, close it.

**Pull request closed without merging**